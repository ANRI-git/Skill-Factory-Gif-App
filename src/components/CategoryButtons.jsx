import React from "react";
import PropTypes from 'prop-types';

export const CategoryButtons = ({ categoriesList, removeCategory }) => {
  return (
    <>
      {categoriesList.map((category, index) => (
          <button className="btn outline-purple" key={index}>
            <div className="d-flex align-items-center">
              <span data-testid="category-button" className="w-100">{category}</span>
              <a className="btn-close flex-shrink-1" onClick={() => removeCategory(index)}></a>
            </div>
          </button>
        ))}
    </>
  );
};

CategoryButtons.propTypes = {
  categoriesList: PropTypes.array.isRequired,
  removeCategory: PropTypes.func.isRequired
}