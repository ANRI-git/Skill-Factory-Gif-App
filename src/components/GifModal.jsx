import React, { useState } from "react";
import Modal from "react-bootstrap/Modal";
import Button from "react-bootstrap/Button";

export const GifModal = ({ title, url }) => {
  const [show, setShow] = useState(false);

  const handleClose = () => setShow(false);
  const handleShow = () => setShow(true);

  return (
    <>
      <Button className="outline-pink" variant="primary" onClick={handleShow}>
        Expand
      </Button>

      <Modal show={show} onHide={handleClose}>
        <Modal.Header closeButton>
          <Modal.Title>{title}</Modal.Title>
        </Modal.Header>
        <Modal.Body className="card-grid">
          <div className="card">
              <img src={url} alt="gif" />
          </div>
        </Modal.Body>
      </Modal>
    </>
  );
};

GifModal.defaultProps = {
  title: 'Gif title',
  url: '...'
}