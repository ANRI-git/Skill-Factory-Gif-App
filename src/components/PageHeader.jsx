import React from "react";

export const PageHeader = () => {
  return (
    <>
      <div className="top-bar text-end">
        <h4>Created by: Andr√©s N. Ramirez</h4>
      </div>
      <div className="header text-center">
        <h1 className="fs-1 mb-5">Gif App</h1>
        <h2 className="fs-6">All the gifs you want on the same place!</h2>
      </div>
    </>
  );
};
