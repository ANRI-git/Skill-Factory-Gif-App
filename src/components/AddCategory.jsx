import React from "react";
import { useState } from "react";
import PropTypes from "prop-types";
import { CategoryButtons } from "./CategoryButtons";

export const AddCategory = ({
  addCategory,
  removeCategory,
  resetCategories,
  categoriesList,
}) => {
  const [inputValue, setInputValue] = useState("");

  const handleInputChange = ({ target }) => {
    setInputValue(target.value);
  };

  const handleSubmit = (event) => {
    event.preventDefault();
    if (inputValue.trim() < 1) return;
    addCategory(inputValue.trim());
    setInputValue("");
  };

  const handleResetCategories = () => {
    setInputValue("");
    resetCategories();
  };

  const handleRemoveCategory = (index) => {
    removeCategory(index);
  };

  return (
    <form aria-label="form" onSubmit={handleSubmit} className="container">
      <div className="row justify-content-center">
        <input
          className="col-12 col-sm-6"
          type="search"
          placeholder="Enter something..."
          value={inputValue}
          onChange={handleInputChange}
        />
        <div className="col-12 col-sm-3 my-2 row justify-content-evenly">
          <button className="col btn outline-pink">Search</button>
          <button
            aria-label="reset"
            className="col btn outline-pink"
            onClick={handleResetCategories}
          >
            Reset
          </button>
        </div>
      </div>
      <div>
        <CategoryButtons
          categoriesList={categoriesList}
          removeCategory={handleRemoveCategory}
        />
      </div>
    </form>
  );
};

AddCategory.propTypes = {
  addCategory: PropTypes.func.isRequired,
  removeCategory: PropTypes.func.isRequired,
  resetCategories: PropTypes.func.isRequired,
  categoriesList: PropTypes.array.isRequired,
};
