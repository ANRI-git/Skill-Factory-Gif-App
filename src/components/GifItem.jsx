import PropTypes from "prop-types";
import { GifModal } from "./GifModal";

export const GifItem = ({ title, url }) => {
  return (
    <div className="card col-12 col-sm-2">
      <img className="card-img" src={url} alt="gif" />
      <p data-testid="test-title">{title}</p>
      <GifModal title={title} url={url} />
    </div>
  );
};

GifItem.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
};
