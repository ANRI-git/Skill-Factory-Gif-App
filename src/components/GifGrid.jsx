import { useFetchGif } from "../hooks/useFetchGif";
import { GifItem } from "./GifItem";

export const GifGrid = ({ category }) => {
  
  const {gifs, loading} = useFetchGif(category);

  return (
    <>
      <h3>Gifs found with '{category}':</h3>
      <div className="card-grid row">
        {loading ? (
            <h2 className="text-center my-5">Loading GIFS...</h2>
        ) : (
          gifs.map((img) => <GifItem key={img.id} {...img} />)
        )}
      </div>
    </>
  );
};
