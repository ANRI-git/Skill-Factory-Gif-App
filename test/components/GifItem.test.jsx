import { render, screen } from "@testing-library/react";
import { GifItem } from "../../src/components/GifItem";

describe("Test en <GifItem />", () => {
  const title = "Dragon Ball";
  const url = "https://dragon-ball.com/goku.jpg";

  test("Debe coincidor con la snapshot", () => {
    // Utilizar para componentes estáticos, nunca dimámicos ya que las snapshots no coincidirían eventualmente
    const { container } = render(<GifItem title={title} url={url} />);

    expect(container).toMatchSnapshot();
  });

  test("Debe de encontrar un texto", () => {
    // Método de prueba pero no utilizado normalmente, se recomienda utilizar 'screen'
    const { getByText } = render(<GifItem title={title} url={url} />);

    expect(getByText(title)).toBeTruthy();
  });

  test("Encontrar por test-id", () => {
    const { getByTestId } = render(<GifItem title={title} url={url} />);

    // Espera que contenga estrictamente solo el texto
    expect(getByTestId("test-title").innerHTML).toBe(title);
    // Espera que contenga el texto de cualquier forma
    expect(getByTestId("test-title").innerHTML).toContain(title);
  });

  test("Debe mostrar la URL y ALT", () => {
    render(<GifItem title={title} url={url} />);

    const { src, alt } = screen.getByRole('img');

    expect(src).toContain(url);
    expect(alt).toContain('gif');
  });
});