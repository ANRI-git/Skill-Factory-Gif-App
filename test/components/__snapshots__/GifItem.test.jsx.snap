// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test en <GifItem /> Debe coincidor con la snapshot 1`] = `
<div>
  <div
    class="card col-12 col-sm-2"
  >
    <img
      alt="gif"
      class="card-img"
      src="https://dragon-ball.com/goku.jpg"
    />
    <p
      data-testid="test-title"
    >
      Dragon Ball
    </p>
    <button
      class="outline-pink btn btn-primary"
      type="button"
    >
      Expand
    </button>
  </div>
</div>
`;
